components/director/internal/tenantmapping/mapper_for_user.go

```diff
@@ -40,26 +46,57 @@ func (m *mapperForUser) GetObjectContext(ctx context.Context, reqData oathkeeper
 
 	ctx = log.ContextWithLogger(ctx, logger)
 
-	log.C(ctx).Info("Getting scopes from groups")
-	scopes = m.getScopesForUserGroups(ctx, reqData)
-	if !hasScopes(scopes) {
-		log.C(ctx).Info("No scopes found from groups, getting user data")
+	if authDetails.Authenticator == nil {
+		log.C(ctx).Info("Getting scopes from groups")
+		scopes = m.getScopesForUserGroups(ctx, reqData)
+		if !hasScopes(scopes) {
+			log.C(ctx).Info("No scopes found from groups, getting user data")
+
+			staticUser, scopes, err = m.getUserData(ctx, reqData, authDetails.AuthID)
+			if err != nil {
+				return ObjectContext{}, errors.Wrapf(err, "while getting user data for user: %s", authDetails.AuthID)
+			}
+		}
+
+		externalTenantID, err = reqData.GetExternalTenantID()
+		if err != nil {
+			if !apperrors.IsKeyDoesNotExist(err) {
+				return ObjectContext{}, errors.Wrapf(err, "could not parse external ID for user: %s", authDetails.AuthID)
+			}
+			log.C(ctx).Warningf("Could not get tenant external id, error: %s", err.Error())
 
-		staticUser, scopes, err = m.getUserData(ctx, reqData, username)
+			log.C(ctx).Info("Could not create tenant context, returning empty context...")
+			return NewObjectContext(TenantContext{}, scopes, authDetails.AuthID, consumer.User), nil
+		}
+	} else {
+		extra, err := json.Marshal(reqData.Body.Extra)
 		if err != nil {
-			return ObjectContext{}, errors.Wrapf(err, "while getting user data for user: %s", username)
+			return ObjectContext{}, err
 		}
-	}
 
-	externalTenantID, err = reqData.GetExternalTenantID()
-	if err != nil {
-		if !apperrors.IsKeyDoesNotExist(err) {
-			return ObjectContext{}, errors.Wrapf(err, "could not parse external ID for user: %s", username)
+		tknAttributes := string(extra)
+
+		authn := authDetails.Authenticator
+		uniqueAttribute := gjson.Get(tknAttributes, authn.Attributes.UniqueAttribute.Key).String()
+		if uniqueAttribute == "" || uniqueAttribute != authn.Attributes.UniqueAttribute.Value {
+			log.C(ctx).Debugf("Request token does not match %q authenticator", authn.Name)
+			return ObjectContext{}, errors.Errorf("unique attribute %q missing or invalid from %s authenticator token - expected [%s], actual [%s]",
+				authn.Attributes.UniqueAttribute.Key, authn.Name, authn.Attributes.UniqueAttribute.Value, uniqueAttribute)
+		}
+
+		log.C(ctx).Info("Getting scopes from token attribute")
+		userScopes, err := reqData.GetUserScopes()
+		if err != nil {
+			return ObjectContext{}, err
+		}
+		scopes = strings.Join(userScopes, " ")
+
+		externalTenantID = gjson.Get(tknAttributes, authn.Attributes.TenantAttribute.Key).String()
+		if externalTenantID == "" {
+			return ObjectContext{}, errors.Errorf("tenant attribute %q missing from %s authenticator token", authn.Attributes.TenantAttribute.Key, authn.Name)
```

Comment:
I would extract the logic of the whole else statement in another implementation of the interface `ObjectContextProvider`. Then instanciate it as it the user mapper currently is instantiated and then provide it to the tenant mapping handler. This will allow for further extending the tenant mapping handler if it accepts map or array of different mappers.

Reply:
Sure thing - sounds like a good idea! Here's the refactoring as you proposed: https://github.com/kyma-incubator/compass/pull/1651/commits/e2b18411e7b42b4feb48f94e0bf08e2b39a5b651

Reply:
Looks nice now!